src/common/log.ts

```diff
@@ -1,27 +1,154 @@
+
+  info(input: string): void {
+    process.stdout.write(grey(input) + '\n');
+  }
+
+  success(input: string): void {
+    process.stdout.write(
+      green(`${StdoutLogger.successPrefix} ${input}`) + '\n'
+    );
+  }
+
+  warn(input: string): void {
+    process.stdout.write(
+      yellow(`${StdoutLogger.warningPrefix} ${input}`) + '\n'
+    );
+  }
+
+  error(input: string): void {
+    process.stdout.write(red(`${StdoutLogger.errorPrefix} ${input}`) + '\n');
+  }
+}
+
+class DummyLogger implements ILogger {
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  info(_input: string): void {}
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  success(_input: string): void {}
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  warn(_input: string): void {}
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  error(_input: string): void {}
+}
+
+export class MockLogger implements ILogger {
+  private stdout: string[];
+  private stderr: string[];
+
+  constructor() {
+    this.stdout = [];
+    this.stderr = [];
+  }
+
+  get stdoutOutput(): string {
+    return this.stdout.join(' ');
+  }
+
+  get stderrOutput(): string {
+    return this.stderr.join(' ');
+  }
+
+  info(input: string): void {
+    this.stdout.push(input);
   }
 
-  let quotedString = '';
-  if (quoted !== undefined && quoted.length !== 0) {
-    quotedString = quoted.map(q => `'${q}'`).join(' ');
+  success(input: string): void {
+    this.stdout.push(input);
   }
 
-  let initialString = (initial ?? '').trim();
-  if (initialString !== '' && quotedString !== '') {
-    initialString = initialString + ' ';
+  warn(input: string): void {
+    this.stderr.push(input);
   }
 
-  return `$ ${envString}${initialString}${quotedString}`;
+  error(input: string): void {
+    this.stderr.push(input);
+  }
+}
+
+/**
+ * Represents logger used in one command lifecycle. For every command run new ILogger instance is created.
+ */
+export class Logger {
+  private static logger: ILogger | undefined = undefined;
+
+  public static setup(quiet?: boolean): void {
+    if (!Logger.logger) {
+      if (!quiet) {
+        Logger.logger = new StdoutLogger();
+      } else {
+        Logger.logger = new DummyLogger();
+      }
+    }
+  }
+
+  /**
+   * Sets up Logger with MockLogger instance, useful in tests of logic where Logger is used.
+   * @returns instance of MockLogger
+   */
+  public static mockLogger(): MockLogger {
+    const mock = new MockLogger();
+    Logger.logger = mock;
+
+    return mock;
+  }
+
+  private static getInstance(): ILogger {
+    if (!Logger.logger) {
+      throw new CLIError('Logger not initialized', { exit: 1 });
+    }
+
+    return Logger.logger;
+  }
+
+  private static getMessage<K extends keyof typeof messages>(
+    messsageTemplate: K,
+    ...args: Parameters<typeof messages[K]>
+  ): string {
+    return (messages[messsageTemplate] as (
```

Comment:
@lukas-valenta @TheEdward162 any idea how to get rid of cast? My goal is to "lock" logger to be usable only with messages. Fist argument id message template - key in messages and second are argument for that template - TS is smart enough to expect arguments matching to message template.

Reply:
i can't seem to get it done :disappointed:

Reply:
This has to be an inference bug though, right? It should work otherwise.

Reply:
more limitation than bug

Reply:
Can I leave it as it is or should I look for some other solutions?

Reply:
leave it for now, i'm trying some stuff